File: lib/cfr/builtins/libs/dart/ui/immutableBuffer.dart
import 'dart:async';
import 'dart:core';
import 'dart:typed_data';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedImmutableBuffer extends VMManagedBox<ImmutableBuffer> {
  VMManagedImmutableBuffer(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['length'] = vmObject.length;
    table['dispose'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.dispose();
      return [];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final ImmutableBuffer vmObject;
}

void loadImmutableBuffer(
    {required HydroState hydroState, required HydroTable table}) {
  table['immutableBufferFromUint8List'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Future>(
          object: ImmutableBuffer.fromUint8List(
              maybeUnBoxAndBuildArgument<Uint8List, dynamic>(
                  luaCallerArguments[1],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<ImmutableBuffer>(boxer: (
      {required ImmutableBuffer vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedImmutableBuffer(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/ui/immutableBuffer.ts
import { IFuture } from "../async/future";
import { IUint8List } from "../typed_data/uint8List";
declare const dart: {
    ui: {
        immutableBufferFromUint8List: (
            list: IUint8List
        ) => IFuture<IImmutableBuffer>;
    };
};
export interface IImmutableBuffer {
    length: number;
    dispose: () => void;
}
export class ImmutableBuffer {
    public readonly length: number = undefined as any;
    public static fromUint8List(list: IUint8List): IFuture<IImmutableBuffer> {
        return dart.ui.immutableBufferFromUint8List(list);
    }
    private readonly _dart_dispose: () => void = undefined as any;
    public dispose(): void {
        return this._dart_dispose();
    }
}
